crates/engine/tree/src/tree/root.rs

```diff
@@ -705,19 +725,42 @@ where
                         trace!(target: "engine::root", "processing StateRootMessage::FinishedStateUpdates");
                         updates_finished = true;
 
-                        let all_proofs_received =
-                            proofs_processed >= updates_received + prefetch_proofs_received;
-                        let no_pending = !self.proof_sequencer.has_pending();
-                        if all_proofs_received && no_pending {
-                            // drop the sender
-                            sparse_trie_tx.take();
-                            debug!(
-                                target: "engine::root",
-                                total_updates = updates_received,
-                                total_proofs = proofs_processed,
-                                "State updates finished and all proofs processed, ending calculation"
-                            );
+                        check_end_condition(CheckEndConditionParams {
```

Comment:
can we make this return bool and take `sparse_trie_tx` within the main loop pls?